integrations/datadog/client.py

```diff
@@ -0,0 +1,270 @@
+
+    This method prepares a URL for use with Datadog webhook integrations.
+    Datadog's webhooks can utilize basic HTTP authentication, requiring credentials embedded within the URL.
+
+    Args:
+        url (str): The original URL.
+        username (str): The username to insert.
+        token (str): The token (likely an API key) to insert.
+
+    Returns:
+        str: The modified URL with inserted credentials, ready for Datadog use.
+
+    Example:
+        new_url = embed_credentials_in_url("https://my.service.example.com", "my_username", "my_api_key")
+        # Use new_url in your Datadog webhook configuration
+    """
+    parsed_url = urlparse(url)
+
+    # Insert credentials into the netloc part of the URL
+    netloc_with_credentials = f"{username}:{token}@{parsed_url.netloc}"
+
+    # Create a new URL with inserted credentials
+    modified_url = urlunparse(
+        (
+            parsed_url.scheme,
+            netloc_with_credentials,
+            parsed_url.path,
+            parsed_url.params,
+            parsed_url.query,
+            parsed_url.fragment,
+        )
+    )
+
+    return modified_url
+
+
+class DatadogClient:
+    def __init__(self, api_url: str, api_key: str, app_key: str):
+        self.api_url = api_url
+        self.dd_api_key = api_key
+        self.dd_app_key = app_key
+
+        self.http_client = http_async_client
+
+    @property
+    async def auth_headers(self) -> dict[str, Any]:
+        return {
+            "DD-API-KEY": self.dd_api_key,
+            "DD-APPLICATION-KEY": self.dd_app_key,
+            "Content-Type": "application/json",
+        }
+
+    async def _send_api_request(
+        self,
+        url: str,
+        params: Optional[dict[str, Any]] = None,
+        json_data: Optional[dict[str, Any]] = None,
+        method: str = "GET",
+    ) -> Any:
+        logger.info(f"Making datadog API {method} request to endpoint {url}")
+
+        response = await self.http_client.request(
+            url=url,
+            method=method,
+            headers=await self.auth_headers,
+            params=params,
+            json=json_data,
+        )
+        response.raise_for_status()
+        return response.json()
+
+    async def get_hosts(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        if cache := event.attributes.get(CacheKeys.HOSTS):
+            logger.info("Picking Datadog Hosts from cache")
+            yield cache
+            return
+
+        start = 0
+        count = MAX_PAGE_SIZE
+
+        while True:
+            url = f"{self.api_url}/api/v1/hosts"
+            result = await self._send_api_request(
+                url, params={"start": start, "count": count}
+            )
+
+            hosts = result.get("host_list")
+            if not hosts:
+                break
+
+            event.attributes.setdefault(CacheKeys.HOSTS, []).extend(hosts)
+            yield hosts
+            start += count
```

Comment:
no need for cache for hosts, as I don't see it is being used in multiple places

Reply:
alright. i have removed it now